core/src/avm1/scope.rs

```diff
@@ -0,0 +1,299 @@
+    /// Scope represents global scope.
+    Global,
+
+    /// Target represents timeline scope. All timeline actions execute with
+    /// the current clip object in lieu of a local scope, and the timeline scope
+    /// can be changed via `tellTarget`.
+    Target,
+
+    /// Scope represents local scope and is inherited when a closure is defined.
+    Local,
+
+    /// Scope represents an object added to the scope chain with `with`.
+    /// It is not inherited when closures are defined.
+    With,
+}
+
+/// Represents a scope chain for an AVM1 activation.
+#[derive(Debug)]
+pub struct Scope<'gc> {
+    parent: Option<GcCell<'gc, Scope<'gc>>>,
+    class: ScopeClass,
+    values: GcCell<'gc, Object<'gc>>,
+}
+
+unsafe impl<'gc> gc_arena::Collect for Scope<'gc> {
+    #[inline]
+    fn trace(&self, cc: gc_arena::CollectionContext) {
+        self.parent.trace(cc);
+        self.values.trace(cc);
+    }
+}
+
+impl<'gc> Scope<'gc> {
+    /// Construct a global scope (one without a parent).
+    pub fn from_global_object(globals: GcCell<'gc, Object<'gc>>) -> Scope<'gc> {
+        Scope {
+            parent: None,
+            class: ScopeClass::Global,
+            values: globals,
+        }
+    }
+
+    /// Construct a child scope of another scope.
+    pub fn new_local_scope(parent: GcCell<'gc, Self>, mc: MutationContext<'gc, '_>) -> Scope<'gc> {
+        Scope {
+            parent: Some(parent),
+            class: ScopeClass::Local,
+            values: GcCell::allocate(mc, Object::bare_object()),
+        }
+    }
+
+    /// Construct a closure scope to be used as the parent of all local scopes
+    /// when invoking a function.
+    pub fn new_closure_scope(
+        mut parent: GcCell<'gc, Self>,
+        mc: MutationContext<'gc, '_>,
+    ) -> GcCell<'gc, Self> {
+        let mut closure_scope_list = Vec::new();
+
+        loop {
+            if parent.read().class != ScopeClass::With {
+                closure_scope_list.push(parent);
+            }
+
+            let grandparent = parent.read().parent;
+            if let Some(grandparent) = grandparent {
+                parent = grandparent;
+            } else {
+                break;
+            }
+        }
+
+        let mut parent_scope = None;
+        for scope in closure_scope_list.iter().rev() {
+            parent_scope = Some(GcCell::allocate(
+                mc,
+                Scope {
+                    parent: parent_scope,
+                    class: scope.read().class,
+                    values: scope.read().values,
+                },
+            ));
+        }
+
+        if let Some(parent_scope) = parent_scope {
+            parent_scope
+        } else {
+            GcCell::allocate(
+                mc,
+                Scope {
+                    parent: None,
+                    class: ScopeClass::Global,
+                    values: GcCell::allocate(mc, Object::bare_object()),
+                },
+            )
+        }
+    }
```

Comment:
My gut feels like we could avoid the `Vec` and `GcCell`s here, since this is basically a clone of the scope chain with a certain scope classes filtered out, if I understand correctly. For example, I think the `Vec` could be avoided by constructing the `Scope` objects in the initial loop and fixing up the parents of the previous one as you go along.

Maybe longer term Scope can use an iterator-like API for the iteration in `resolve` etc., and these just could ignore certain `ScopeClass`es as appropriate while iterating. Then the scope chains will only be appended to and never have to be copied?

Maybe eventually we can just add on the the

Maybe instead

Reply:
The iterator approach seems difficult, because we might have a function defined in a `with` block that calls another function with another with block, etc. So it wouldn't be as simple as "resolve without with blocks", since we only want those defined in the current function.

Removing the extra allocations is an easy perf win though.